### ISSUE TITLE
Proposal for Function Overloading in Modelica 3.2

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#207]
**Reported by svorkoetter on 11 Sep 2009 14:59 UTC**
At the 62nd design meeting, a proposal for overloading infix operators was approved, but the corresponding proposal for overloading functions was not. One issue left unresolved was how to define overloads for functions that may be part of another package. For example, an implementor of a Complex data type may wish to make the Modelica.Math.sin function available for Complex values. With the original proposal, this would have required modifying the Modelica.Math package itself, which is clearly not acceptable.

I propose that we extend the mechanism used for overloading infix operators. What is needed is a way to describe the package whose function is to be overloaded. The existing "within" keyword seems like a natural fit since that is very close to what it already means, and it currently is allowed only at the top level (outside of any class definition) and thus will not introduce any backward incompatibilities.

In the context of the definition of Complex given in section 14.6 of the Modelica 3.1 specification, this would look as follows:
```text/x-modelica
record Complex "Record defining a Complex number"
   Real re "Real part of Complex number";
   Real im "Imaginary part of Complex number";
   ...

   operator function sin
      within Modelica.Math;
      input Complex x;
      output Complex result "= sin(x)";
      import Math = Modelica.Math;
   algorithm
      result := Complex(Math.sin(x.re)*Math.cosh(x.im),
                        Math.cos(x.re)*Math.sinh(x.im));
   end sin;

   ...
end Complex;
```
The extension here from the Modelica 3.1 spec is just addition of the clause, "within name", at the top of the operator function definition. The proposed semantics are:

The addition of "within packagename" in an operator function definition has the effect of adding the function to the specified package in addition to it being added as a method in the current record.

The above definition of Complex can be used as follows:
```text/x-modelica
Complex j = Complex.j();
Complex c1 = 2 + 3*j;

Complex s1 = Complex.sin(c1);       // explicit call
Complex s2 = c1.sin();              // call via instance
Complex s3 = Modelica.Math.sin(c1); // overloaded call
```

Each of the last three are functionally equivalent.

The only change required in the grammar is in the "class_specifier" production:

```
class_specifier :
    IDENT [ within name ";" ] string_comment composition end IDENT
    ...
```
This is subject to the following semantic restrictions:
 1. The optional "within name ;" sequence can only appear directly within an operator definition or a short-hand operator function definition (ie. "operator function x within a.b.c; ..." is equivalent to "operator x within a.b.c; function x1 ...").
 2. The name must refer to an existing package, and standard scoping rules for package names are used to resolve it (ie. the same rules that would apply to import name in the same context).
 3. The name of the operator being defined must correspond to the name of an existing function within the referenced package.
 4. No calling signature of the operator being defined can correspond to the calling signature of that function (ie. this mechanism cannot be used to replace an existing function outright).

Determination of when to call an overloaded function is as per section 14.4 of the Modelica 3.1 specification, extended in the obvious way to n-ary functions.

<sub>The [attachment:ModelicaFunctionOverloading.doc attached .doc file] contains the same text as above, just more nicely formatted.</sub>

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/207

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 19 Nov 2009 15:35 UTC**
## From [milestone:Design63]
Discussion on function overloading. Currently three proposals:
 a. Similar as operator overloading (advantage: user can inspect at one place the overloaded functions; drawback: only library developed can include new overloaded functions).
 b. Proposal from Stefan sent out before the meeting (advantage: user can add overloaded functions without modifying the library where the function to be overloaded is present; drawback: not easy to get an overview which overloaded functions are present).
 c. Use restricted Ada/Koening lookup (poles(tf) -> TypeOf(tf).poles(tf)) (advantage: short names)
Discussion will continue. Prototype implementations would be useful.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 25 Nov 2009 12:09 UTC**
This was discussed at [milestone:Design63] and the outcome were three proposals:
 1. Similar as operator overloading, i.e., the different functions are defined at one point
 2. Proposal from Stefan so that functions are defined somewhere else and define what to overload
 3. Use restricted Ada/Koening lookup
 4. Use member function syntax.

Hilding, Hans and myself (Martin) have the same opinion:
 * (1) is bad
 * (2) is also bad (since an existing library is modified from the outside, without control of the library developer)
 * (3) seems reasonable
 * (4) seems to be a non-standard hack

So, we would prefer (3).

Peter (F.), Michael, Gerd, Sébastien ..., what is your opinion. If there seems to be agreement to use (3), one could work out the details and make a prototype implementation until the next meeting.

One issue is that we would like to include "record Complex" in the next version of MSL, in order that in follow-up versions of MSL, libraries building upon Complex can be provided.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 25 Nov 2009 12:56 UTC**
My preference (especially in light of the followup email lowering the priority for this feature) would be 4.  The reason I prefer this is that if we eventually adopt some kind of feature then it will be possible to do things like:

```text/x-modelica
record Complex "Record defining a Complex number"
   Real re "Real part of Complex number";
   Real im "Imaginary part of Complex number";
   function ReturnsComplex sin = function Methods.sin(re=re,im=im);
   function ReturnsReal mag = function Methods.mag(re=re,im=im);
   function ReturnsComplex j = function Complex(re=0.0, im=1.0); // Record constructor is already a function
protected
   package Methods
     partial function ReturnsComplex
       output Complex result;
     end ReturnsComplex;
     partial function ReturnsReal
       output Real result;
     end ReturnsReal;
     function mag
       input Real re;
       input Real im;
       output Real result;
     algorithm
       result := sqrt(re^2+im^2);
     end mag;
     function sin
       input Real re;
       input Real im;
       output Complex result "= sin(x)";
       import Math = Modelica.Math;
     algorithm
        result := Complex(Math.sin(re)*Math.cosh(im),
                          Math.cos(re)*Math.sinh(im));
     end sin;
   end Methods;
   ...
end Complex;
```

So there are two issues.  First, we will spend time developing two different proposals when only one is required.  Second, in the future when it is possible to express closures there will be two ways of accomplishing the same thing which is also just bad language design.  This fits very naturally into the idea of closures and would be a natural extension of the "functions as arguments to functions" semantics we are already working on in parallel.  Their use would be identical to the previous examples except for the overloaded call, i.e.,

```text/x-modelica
Complex j = Complex.j();
Complex c1 = 2 + 3*j;

Complex s1 = Complex.sin(c1);       // explicit call
Complex s2 = c1.sin();              // call via instance
```

Frankly, I don't really like the idea of the overloaded call anyway.  It seems very strange that there should be a function in Modelica.Math called sin that takes a complex argument but then one has to scour the entire package hierarchy in an effort to find it (and rather shocking to beginners that it isn't in Modelica.Math).  I don't see why that usage is even significant.  The above examples seem quite sufficient to me.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 25 Nov 2009 14:15 UTC**
Somehow the previous comment got messed up.  The first paragraph should say "adopt some kind of closure feature".  Sorry about the confusion.

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 25 Nov 2009 15:07 UTC**
Replying to [comment:2 dietmarw]:
>  2. Proposal from Stefan so that functions are defined somewhere else and define what to overload
>  ...
> Hilding, Hans and myself (Martin) have the same opinion:
>  \* (1) is bad
>  \* (2) is also bad (since an existing library is modified from the outside, without control of the library developer)

I disagree that (2) is bad. One is not modifying an existing library, since no 
existing functionality of the library is changed. (i.e. one can't change the 
behaviour of Math.sin for Reals, and one can't add functions to it either). One is 
simply extending the library to work with new types (that it would not have 
worked with before anyway).

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 25 Nov 2009 15:10 UTC**
Replying to [comment:3 mtiller]:
> It seems very strange that there should be a function in Modelica.Math called sin that takes a complex argument but then one has to scour the entire package hierarchy in an effort to find it (and rather shocking to beginners that it isn't in Modelica.Math).

I think that to some extent that is a tool issue. A tool that is designed for editing Modelica libraries (or more generally, working with and in an environment containing several Modelica libraries), should be able to show you all the implementations of Modelica.Math.sin. The concept of having to scour (aka grep) a bunch of source files to find a scattered implementation is probably a bit dated.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 25 Nov 2009 15:29 UTC**
While it is true that a tool can help to find the definition of a function that "appears" to be in Modelica.Math even though it's definition is elsewhere, the larger point here (for me) is whether such usage (i.e. the ability to "mount" new definitions within an existing package hierarchy) should be a requirement.  So, while a tool may be able to help project a more uniform perspective, I don't see why the ability to "mount" a function in an existing package hierarchy should be a requirement (as it relates to my alternative, closure based approach).

**Comment by modelica-trac-importer** [comment]
**Comment by svorkoetter on 25 Nov 2009 15:37 UTC**
The thing I don't like about having to explicitly qualify a call with either the class or instance of the new object is that you lose uniformity with the built-in types, and uniformity with the overloading of operators (which to me are just built-in functions with syntactic sugar on top). The example I've given in the past is changing the declaration of a variable x from Real to Complex, where with your proposal, you'd then have to go change all the calls to sin(x) as well.

I like uniformity in my programming languages. :-)

**Comment by modelica-trac-importer** [comment]
**Comment by magnus on 25 Nov 2009 16:30 UTC**

I disagree that this is a tool issue. My opinion is that every Modelica source file should be readable as is, without the aid of tools. In general, I think that an overall goal of the language design must be to avoid strangeness and inconsistencies. The tool should not be made responsible to cover up for strange language design. 

I fully agree with with Mike that for a beginner, or even normal, user it would be very strange to see a statement like `Modelica.Math.sin(complex)`, when the actual function that will be called is a completely different one.

Out of the four proposals I am leaning towards (3) "restricted Ada/Koening lookup", although I am not sure I understand the details, or (4). With (3), I assume that `TypeOf(tf).poles(tf` would expand to something like `ComplexPackage.Complex` that would be a record where the sin function is defined? Is there a code example somewhere?

Proposal (4) might be the best alternative, but then member functions should be introduced first, which I think they should. Would Complex then simply be a normal object with a sin method? Called as c.sin() or sin(c)? Would the method also have access to the object instance scope, or must elements in the object instance be bound explicitly to input arguments, like in Mikes example above? And would overloading of methods be supported for derived classes? If so, with redeclare syntax? I think the syntax for declaration of methods in Mikes example is too difficult.

Sorry for jumping in to the discussion like this.

Replying to [comment:6 anonymous]:
> Replying to [comment:3 mtiller]:
> > It seems very strange that there should be a function in Modelica.Math called sin that takes a complex argument but then one has to scour the entire package hierarchy in an effort to find it (and rather shocking to beginners that it isn't in Modelica.Math).
> 
> I think that to some extent that is a tool issue. A tool that is designed for editing Modelica libraries (or more generally, working with and in an environment containing several Modelica libraries), should be able to show you all the implementations of Modelica.Math.sin. The concept of having to scour (aka grep) a bunch of source files to find a scattered implementation is probably a bit dated.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 25 Nov 2009 16:44 UTC**
> Proposal (4) might be the best alternative, but then member functions should be introduced
> first, which I think they should. Would Complex then simply be a normal object with a sin
> method? Called as c.sin() or sin(c)? Would the method also have access to the object instance 
> scope, or must elements in the object instance be bound explicitly to input arguments, like in
> Mikes example above? And would overloading of methods be supported for derived classes? If so,
> with redeclare syntax? I think the syntax for declaration of methods in Mikes example is too
> difficult.

Magnus, I'll return the favor and agree with you.  The syntax is a bit cumbersome, I agree.  But the point I wanted to make is that we have two proposals in parallel where I think we should simply focus on one (closures).

You raised two issue, let me address each of them.

First, you mention overriding of the methods.  With closures this is simple because a closure is a value, not a type.  So you just change the value, e.g.

```
record ReallyComplex
  extends Complex(sin = function ReallyComplex.sin(re=re,im=im));
end ReallyComplex;
```

Now, as I said, the syntax is a bit ugly.  I see this as a stepping stone to nicer syntax but we don't have to get all the way there to have something useful.  For example, in the future we might introduce some improved syntax specifically for methods (I'm not even going to provide an example because then we'll end up discussing that which I don't think is necessary right now).  But it could just be simple syntactic sugar over what ultimately resulted in a closure.

In conclusion, why not extend the "functions as arguments to functions" proposal to allow "function as elements in records" (which doesn't require any real leap) and use those as methods for now.  Later, we define some nicer syntax for defining them but we don't need to redefine the semantics.

P.S. - The method would be invoked as c.sin().  Access to instance variables would have to be allowed (I think you told me Dymola is already supporting this in some contexts?!?).

**Comment by modelica-trac-importer** [comment]
**Comment by magnus on 25 Nov 2009 17:06 UTC**
I like uniformity too.

But I guess that methods **will** be introduced in Modelica some way or another. If the calling syntax for a method will be `instance.method(arguments)`, then this will be a uniform pattern. (One may, or perhaps should, also allow `method(instance,arguments)`, but that might introduce problems in distinguishing a method and a function with the same name in the lookup?) I don't think that this is inconsistent with the functions on built-in types, like Modelica.Math.sin(x), since methods and functions are two separate things.

Mixed comments:
  * I think that sin(x) is a nicer "engineering syntax" than x.sin()
  * It is inconvenient to have to change calling syntax when changing from a built in type to a user defined class (sin(x) -> x.sin())
  * In cases 3. and 4. it is no longer overloading, but another lookup scheme.
  * The example below demonstrates (I think) that a method based solution would not necessarily mean that calling syntax would have to change (only the import and type statements):

```
// (i) original code with built in type
import Modelica.Math as M;
type T = Real;
T c;
T s = M.sin(c);

// (ii) changed code with class and methods
import Modelica.Math.Complex as M;
type T = Modelica.Math.Complex;
T c;
T s = M.sin(c); // use function syntax when method is called statically
T t = c.sin(); // use method syntax on instance
```



Replying to [comment:8 svorkoetter@…]:
> The thing I don't like about having to explicitly qualify a call with either the class or instance of the new object is that you lose uniformity with the built-in types, and uniformity with the overloading of operators (which to me are just built-in functions with syntactic sugar on top). The example I've given in the past is changing the declaration of a variable x from Real to Complex, where with your proposal, you'd then have to go change all the calls to sin(x) as well.
> 
> I like uniformity in my programming languages. :-)

**Comment by modelica-trac-importer** [comment]
**Comment by magnus on 25 Nov 2009 17:22 UTC**
Replying to [comment:10 mtiller]:
> In conclusion, why not extend the "functions as arguments to functions" proposal to allow "function as elements in records" (which doesn't require any real leap) and use those as methods for now. 

Good idea.

> P.S. - The method would be invoked as c.sin().  Access to instance variables would have to be allowed (I think you told me Dymola is already supporting this in some contexts?!?).

Dymola is supporting binding arguments to instance variables, like
```
model M
  extends M0(redeclare function f = f0(x=x));
  Real x;
end M;
```
but not access to the instance namespace. Access to the instance namespace is, I guess, a quite complex issue, and perhaps not needed. Perhaps it would be sufficient to allow binding in line with what Dymola already allow (again, rough syntax):
```
model M
  Real x;
  method g = f(x=x);
  function f
    input Real x;
    input Real y;
    ...
  end f;
  ...
end M;
```
Probably one would also like to have 'self' as an implicit (input) argument to a method. Having both binding like in the example and the self keyword would allow for easily reusing functions as methods (as in the example), but also writing more "object-oriented" methods with access to instance namespace through self. A static call `M.g(m)` (as in my previous post) would require the self argument to be given explicitly.

**Comment by modelica-trac-importer** [comment]
**Comment by sebastien on 26 Nov 2009 13:45 UTC**
Replying to [comment:2 dietmarw]:
> Peter (F.), Michael, Gerd, Sébastien ..., what is your opinion. If there seems to be agreement to use (3), one could work out the details and make a prototype implementation until the next meeting.

 I'm completely lost, sorry. I don't understand why we have to consider the special case of "function overloading". Operator and function overloading is basically the same thing (differences are mainly at the syntax level).
 Whatever the selected solution, it is very important to isolate those features from the rest of the language in the hope that a cleaner solution would be designed in the future (as we said during the last design meeting, the current proposal for operator overloading has to be considered as a patch against Modelica 3.1).

**Comment by modelica-trac-importer** [comment]
**Comment by Stefan Vorkoetter on 26 Nov 2009 14:21 UTC**
Replying to [comment:13 sebastien]:
> I don't understand why we have to consider the special case of "function overloading". Operator and function overloading is basically the same thing (differences are mainly at the syntax level).

Thank you Sébastien! This is exactly the point I'm trying to make, and which was the crux of my original proposal. We already have **have** operator overloading, so we should make function overloading work in the analogous way.

Perhaps I might be accused of just wanting to keep the compiler simpler, but my experience has been that clean compilers are a sign of a clean language.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 15 Mar 2010 11:38 UTC**
For timing reasons it is not possible to discuss and include this in Modelica 3.2. Since at the 66th meeting we concentrate only on Modelica 3.2, this topic is shifted to the follow-up meeting.

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 15 Sep 2010 15:56 UTC**
Might be included in Modelica 4 - at least part of the discussion.
